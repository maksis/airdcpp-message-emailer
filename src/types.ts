export type Message = {
  from: { nick: string };
  text: string;
  time: number; // seconds epoch
};

export type SessionId = string | number;

export type SessionInfoGetter = (id: SessionId) => Promise<SessionInfo>;

export type SessionBase = {
  // id: SessionId;
  message_counts: {
    unread: {
      bot: number;
      status: number;
      user: number;
    };
  };
};

export type HubSession = SessionBase & {
  identity: {
    name: string;
  };
  user?: {
    nicks?: string;
  };
};

export type PrivateChatSession = SessionBase & {
  user: {
    nicks: string;
  };
  identity?: {
    name?: string;
  };
};

export type SessionInfo = HubSession | PrivateChatSession;

export type SessionNameGetter = (sessionInfo: SessionInfo) => string;

export type SessionType = {
  title: string;
  sessionNameGetter: SessionNameGetter;
};

export type MessageCache = Record<string, Message[]>;
