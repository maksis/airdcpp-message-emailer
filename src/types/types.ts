import { ChatMessage, SessionBase, SessionId, SessionInfo } from "./api";

export type SessionInfoFetcher<SessionType extends SessionBase> = (id: SessionId, skipCache?: boolean) => Promise<SessionType>;
export type SessionNameGetter = (sessionInfo: SessionInfo) => string;

export type SessionParser = {
  title: string;
  sessionNameGetter: SessionNameGetter;
};

export type MessageCache = Record<string, ChatMessage[]>;
