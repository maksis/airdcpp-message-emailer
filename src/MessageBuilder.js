
const reduceMessageSummary = (reduced, message) => {
	const timeOffset = (-1) * new Date().getTimezoneOffset() * 60 * 1000;
	const timeString = new Date((message.time * 1e3) + timeOffset).toISOString().slice(-13, -5);
	reduced += `[${timeString}] <${message.from.nick}> ${message.text}\n`;
	return reduced;
};

const hasUnread = (sessionInfo) => {
	const { unread } = sessionInfo.message_counts;
	return unread.bot > 0 || unread.user > 0;
};

const constructSummary = async ({ title, sessionNameGetter }, cache, sessionInfoGetter) => {
	let sessionsInfos = (await Promise.all(Object.keys(cache)
		.map(sessionInfoGetter)))
		.filter(hasUnread);

	if (sessionsInfos.length == 0) {
		return '';
	}

	let ret = '';
	ret += `------- ${title} -------`;
	ret += '\n\n';

	return sessionsInfos.reduce((reduced, sessionInfo) => {
		reduced += `-- ${sessionNameGetter(sessionInfo)} --`;
		reduced += '\n\n';
		
		reduced += cache[sessionInfo.id].reduce(reduceMessageSummary, '');
		reduced += '\n\n';

		return reduced;
	}, ret);
}

export default {
	constructSummary,
};