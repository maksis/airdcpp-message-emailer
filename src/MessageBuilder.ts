import { Message, SessionType, MessageCache, SessionInfoGetter, SessionInfo, SessionId } from './types';

const formatTimeStamp = (time: number) => {
  const timeOffset = (-1) * new Date().getTimezoneOffset() * 60 * 1000;
  return new Date((time * 1e3) + timeOffset).toISOString().slice(-13, -5);
}

const reduceMessageSummary = (reduced: string, message: Message) => {
  const timeString = formatTimeStamp(message.time);
  reduced += `[${timeString}] <${message.from.nick}> ${message.text}\n`;
  return reduced;
};

const hasUnread = (sessionInfo: SessionInfo) => {
  const { unread } = sessionInfo.message_counts;
  return unread.bot > 0 || unread.user > 0;
};

const mapSessionInfo = async (sessionId: SessionId, sessionInfoGetter: SessionInfoGetter) => {
  try {
    return await sessionInfoGetter(sessionId);
  } catch (e) {
    // Session removed, assume that is has been read
  }

  return null;
};

const constructSummary = async ({ title, sessionNameGetter }: SessionType, cache: MessageCache, sessionInfoGetter: SessionInfoGetter) => {
  if (!cache) {
    return '';
  }

  const sessionsInfos = (await Promise.all(Object.keys(cache)
    .map(sessionId => mapSessionInfo(sessionId, sessionInfoGetter))))
    .filter(sessionInfo => !!sessionInfo)
    .filter(hasUnread);

  if (sessionsInfos.length == 0) {
    return '';
  }

  let ret = '';
  ret += `------- ${title} -------`;
  ret += '\n\n';

  return sessionsInfos.reduce((reduced: string, sessionInfo: any) => {
    reduced += `-- ${sessionNameGetter(sessionInfo)} --`;
    reduced += '\n\n';
    
    reduced += cache[sessionInfo.id].reduce(reduceMessageSummary, '');
    reduced += '\n\n';

    return reduced;
  }, ret);
}

export default {
  constructSummary,
};
