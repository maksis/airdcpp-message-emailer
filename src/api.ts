import { APISocket } from 'airdcpp-apisocket';
import { SessionInfoFetcher } from './types/types';
import { HubSession, PrivateChatSession, SeverityEnum } from './types';


export const API = (socket: APISocket) => {
  const getPrivateChatSession: SessionInfoFetcher<PrivateChatSession> = (sessionId) => socket.get(`private_chat/${sessionId}`);
  const getHubSession: SessionInfoFetcher<HubSession> = (sessionId) => socket.get(`hubs/${sessionId}`);


  const postEvent = async (text: string, severity: SeverityEnum) => {
    return socket.post(
      'events',
      {
        text,
        severity,
      }
    );
  };

  return {
    getPrivateChatSession,
    getHubSession,
    postEvent,
  }
};

export type APIType = ReturnType<typeof API>;
