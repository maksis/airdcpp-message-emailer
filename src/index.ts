import 'source-map-support/register';
import { ManagedExtension } from 'airdcpp-extension';
import init from './main';

// Entry point that is executed by the extension manager
//
// The file isn't executed when running development server so it shouldn't 
// generally contain any extension-specific code

// See https://github.com/airdcpp-web/airdcpp-extension-js for usage information
ManagedExtension(init, {
  // Possible custom options for airdcpp-apisocket can be listed here
});
