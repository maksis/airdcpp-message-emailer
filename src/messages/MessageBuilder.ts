import { ChatMessage, SessionParser, MessageCache, SessionInfoFetcher, SessionInfo, SessionId } from '../types';

const formatTimeStamp = (time: number) => {
  const timeOffset = (-1) * new Date().getTimezoneOffset() * 60 * 1000;
  return new Date((time * 1e3) + timeOffset).toISOString().slice(-13, -5);
}

const reduceMessageSummary = (reduced: string, message: ChatMessage) => {
  const timeString = formatTimeStamp(message.time);
  reduced += `[${timeString}] <${message.from.nick}> ${message.text}\n`;
  return reduced;
};

const hasUnread = (sessionInfo: SessionInfo) => {
  const { unread } = sessionInfo.message_counts;
  return unread.bot > 0 || unread.user > 0;
};

const mapSessionInfo = async (sessionId: SessionId, SessionInfoFetcher: SessionInfoFetcher<SessionInfo>) => {
  try {
    // Skip the session cache to get the latest read status
    const session = await SessionInfoFetcher(sessionId, true);
    return session;
  } catch (e) {
    // Session removed, assume that is has been read
  }

  return null;
};

const constructSummary = async ({ title, sessionNameGetter }: SessionParser, cache: MessageCache, SessionInfoFetcher: SessionInfoFetcher<SessionInfo>) => {
  if (!cache) {
    return '';
  }

  const sessionsInfos = (await Promise.all(Object.keys(cache)
    .map(sessionId => mapSessionInfo(sessionId, SessionInfoFetcher))))
    .filter(sessionInfo => !!sessionInfo)
    .filter(hasUnread);

  if (sessionsInfos.length == 0) {
    return '';
  }

  let ret = '';
  ret += `------- ${title} -------`;
  ret += '\n\n';

  return sessionsInfos.reduce((reduced: string, sessionInfo: any) => {
    reduced += `-- ${sessionNameGetter(sessionInfo)} --`;
    reduced += '\n\n';
    
    reduced += cache[sessionInfo.id].reduce(reduceMessageSummary, '');
    reduced += '\n\n';

    return reduced;
  }, ret);
}

export default {
  constructSummary,
};
