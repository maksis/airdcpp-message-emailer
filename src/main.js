'use strict';

import nodemailer from 'nodemailer';

import MessageBuilder from './MessageBuilder';
import SessionTypes from './SessionTypes';

import { DEFAULT_TO, SettingDefinitions } from './SettingDefinitions';
const CONFIG_VERSION = 1;

const SettingsManager = require('airdcpp-extension-settings');

export default function (socket, extension) {
	const settings = SettingsManager(socket, {
		extensionName: extension.name, 
		configFile: extension.configPath + 'config.json',
		configVersion: CONFIG_VERSION,
		definitions: SettingDefinitions,
	});

	let transporter;
	let flushTimeout;
	let hostname;

	const cache = {
		privateMessages: {},
		hubMessages: {},
	};

	const hasConfig = () => settings.getValue('to') !== DEFAULT_TO;

	const constructSmtpSettings = () => {
		const ret = {
			host: settings.getValue('smtp_host'),
			port: settings.getValue('smtp_port'),
			secure: settings.getValue('smtp_secure'),
		}

		const user = settings.getValue('smtp_user');
		const pass = settings.getValue('smtp_password');
		if (user && pass) {
			ret['auth'] = {
				user,
				pass,
			}
		}

		return ret;
	};

	const constructMail = (text) => {
		return {
				from: settings.getValue('from'),
				to: settings.getValue('to'),
				subject: `Chat summary from AirDC++ (${hostname})`,
				text,
		};
	};

	const privateChatInfoGetter = sessionId => socket.get(`private_chat/${sessionId}`);
	const hubInfoGetter = sessionId => socket.get(`hubs/${sessionId}`);

	// Send cached messages (if there are any)
	const flushCache = async () => {
		socket.logger.verbose('Flushing cache...');
		let messageSummary = '';
		messageSummary += await MessageBuilder.constructSummary(SessionTypes.privateChat, cache.privateMessages, privateChatInfoGetter);
		messageSummary += await MessageBuilder.constructSummary(SessionTypes.hub, cache.hubMessages, hubInfoGetter);

		if (messageSummary) {
			transporter.sendMail(constructMail(messageSummary), (error, info) => {
				if (error) {
					socket.logger.error(`Failed to send email: ${error}`);
					socket.post('events', {
						text: `${extension.name}: failed to email message summary (see the extension error log for more information)`,
						severity: 'error',
					});
				} else {
					socket.logger.info(`Message ${info.messageId} sent: ${info.response}`);
				}
			});
		} else {
			socket.logger.verbose('Nothing to send');
		}

		cache.privateMessages = {};
		cache.hubMessages = {};
		flushTimeout = undefined;
	};

	const onIncomingMessage = (cacheKey, message, sessionId) => {
		if (!hasConfig() || !settings.getValue('flush_interval')) {
			socket.logger.verbose('Caching disabled due to current configuration');
			return;
		}
		
		const sessionCache = cache[cacheKey];
		sessionCache[sessionId] = sessionCache[sessionId] || [];
		sessionCache[sessionId].push(message);

		if (!flushTimeout) {
			const flushMs = settings.getValue('flush_interval') * 60 * 1000;
			socket.logger.verbose('Scheduling flush', flushMs);
			flushTimeout = setTimeout(flushCache, flushMs);
		}
	};

	const isSmtpSetting = key => key.indexOf('smtp_') === 0;

	extension.onStart = async (sessionInfo) => {
		hostname = sessionInfo.system_info.hostname;
		settings.onValuesUpdated = updatedValues => {
			if (Object.keys(updatedValues).find(isSmtpSetting)) {
				socket.logger.verbose('Creating mailer interface');
				try {
					transporter = nodemailer.createTransport(constructSmtpSettings());
				} catch (e) {
					socket.logger.error(`Failed to initialize mailer interface: ${e}`);
					process.exit(1);
				}

				if (hasConfig()) {
					transporter.verify((error, success) => {
						socket.logger.verbose('Verifying settings');
						if (error) {
							socket.post('events', {
								text: `${extension.name}: SMTP setting validation failed (${error})`,
								severity: 'error',
							});
						} else {
							socket.logger.verbose('Setting validation succeeded');
							/*socket.post('events', {
								text: `${extension.name}: SMTP settings were validated successfully`,
								severity: 'info',
							});*/
						}
					});
				}
			}
		};

		await settings.load();
		socket.logger.verbose('Settings were loaded');

		if (settings.getValue('send_hub_logs')) {
			socket.addListener('hubs', 'hub_message', onIncomingMessage.bind(this, 'hubMessages'));
		}

		if (settings.getValue('send_private_logs')) {
			socket.addListener('private_chat', 'private_chat_message', onIncomingMessage.bind(this, 'privateMessages'));
		}
	};

	extension.onStop = () => {
		clearTimeout(flushTimeout);
	};
};