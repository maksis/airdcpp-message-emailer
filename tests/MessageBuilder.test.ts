import MessageBuilder from '../src/messages/MessageBuilder';
import { SessionPropertyParsers } from '../src/messages/SessionPropertyParsers';
import { SessionId } from '../src/types';

import { describe, test, expect, beforeEach, afterEach } from 'vitest';
import { HubUser1, HubUser2, MessageCountsRead, PrivateChat1, PrivateChat2, User1Message1, User1Message2, User2Message1, User2Message2 } from './mocks/mock-entities';


describe('MessageBuilder', () => {
  const getTimezoneOffset = Date.prototype.getTimezoneOffset;

  beforeEach(() => {
    Date.prototype.getTimezoneOffset = () => {
      return -180;
    }
  });

  afterEach(() => {
    Date.prototype.getTimezoneOffset = getTimezoneOffset;
  });

  const cache: Record<string, any[]> = {
    [HubUser1.cid]: [
      User1Message1,
      User1Message2,
    ],
    [HubUser2.cid]: [
      User2Message1,
      User2Message2,
    ]
  };

  test('should format messages', async () => {
    const SessionInfoFetcher = (sessionId: SessionId) => {
      return Promise.resolve(sessionId === PrivateChat1.id ? PrivateChat1 : PrivateChat2);
    };

    const summary = await MessageBuilder.constructSummary(SessionPropertyParsers.privateChat, cache, SessionInfoFetcher);
    expect(summary).toMatchSnapshot();
  });

  test('should skip read sessions', async () => {
    const SessionInfoFetcher = (_sessionId: SessionId) => Promise.resolve({
      ...PrivateChat1,
      message_counts: MessageCountsRead
    });

    const summary = await MessageBuilder.constructSummary(SessionPropertyParsers.privateChat, cache, SessionInfoFetcher);
    expect(summary).toEqual('');
  });

  test('should handle removed sessions', async () => {
    const SessionInfoFetcher = (_sessionId: SessionId) => Promise.reject('Session removed');

    const summary = await MessageBuilder.constructSummary(SessionPropertyParsers.privateChat, cache, SessionInfoFetcher);
    expect(summary).toEqual('');
  });
});
