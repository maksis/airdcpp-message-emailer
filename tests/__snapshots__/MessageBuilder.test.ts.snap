// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MessageBuilder > should format messages 1`] = `
"------- Private messages -------

-- [100]User1 --

[19:34:42] <[100]User1> User 1 message 1
[19:43:02] <[100]User1> User 1 message 2


-- [100]User2 --

[19:51:22] <[100]User2> User 2 message 1
[19:59:42] <[100]User2> User 2 message 2


"
`;
